Eeprom make table const or remove

KurtE opened the pull request:
THIS IS AN ALTERNATIVE TO PR: #58 

Use this one or the other. 

With this change I did what I mentioned in the issue: #54 
Where I could define the #define VIRTADDVARTAB(index)   (index)

This one started off from what was in PR: #58  and added this in, where there is a #define that can either use the const table or does not build the const table at all...

With using the table my test program compiled: size: 58780 with RAM of 12372, 
With the option (as in the PR), the code shrank to: 57724 ram the same.

KurtE commented:
Just an FYI: If I run a simple test like:
```
#include <EEPROM.h>
void setup() {
  // put your setup code here, to run once:
  while (!Serial) ;
  pinMode(LED_BUILTIN, OUTPUT);
  Serial.begin(115200);
  Serial.println("EEPROM Test");
  // lets try writing to EEPROM
  Serial.printf("EEPROM length: %d\n", EEPROM.length());
  uint16_t max_index = EEPROM.length();
  if (max_index > 32) max_index = 32;
  uint32_t start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    EEPROM.write(i, i & 0xff);
  }
  Serial.printf("Write time: %d\n", micros() - start_time);
  Serial.println("Start read back");
  start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    uint8_t b = EEPROM.read(i);
    if (b != (i & 0xff)) {
      Serial.printf("Read mismatch: %d %x!=%x\n", i, b, i & 0xff);
    }
  }
  Serial.printf("Read time: %d\n", micros() - start_time);
}

void loop() {
  digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
  delay(250);
 }
```
Where I did not restrict the writes and tried to do EEPROM.write(i, i&0xff) for the full 512 bytes, the 
system hung out to dry... I had to do a firmware recovery on OpenCM9.... 

I believe I had the same issue on released version as well.    It works on both restricted to 32... Will now walk it up to 128 and first try on released...

KurtE commented:
Quick update on test program.  It is not just the count, but the cumulative count of the EEPROM writes. 
That is if I do the test Up to 64, I can run the test I believe 4 times, but on the 5th time it hangs.  My guess is that there is a bug in either the update to 2nd page or wrapping around... 

May need to open other issue, as this testing was done on the current release code

OpusK commented:
@KurtE 

I think this PR is better than #58 because it is a little more optional.
It would be nice if you could go on here!

KurtE commented:
@OpusK 
I changed the value of PAGE1 from 0x40 to 1, 
Running test sketch:
```
#define MAX_INDEX 256
#include <EEPROM.h>
void setup() {
  // put your setup code here, to run once:
  while (!Serial) ;
  pinMode(LED_BUILTIN, OUTPUT);
  Serial.begin(115200);
  Serial.println("EEPROM Test");
  // lets try writing to EEPROM
  uint16_t max_index = EEPROM.length();
  if (max_index > MAX_INDEX) max_index = MAX_INDEX;
  Serial.printf("EEPROM length: %d Max Test: %d\n", EEPROM.length(), max_index);
  uint32_t start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    EEPROM.write(i, i & 0xff);
  }
  Serial.printf("Write time: %d\n", micros() - start_time);
  Serial.println("Start read back");
  start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    uint8_t b = EEPROM.read(i);
    if (b != (i & 0xff)) {
      Serial.printf("Read mismatch: %d %x!=%x\n", i, b, i & 0xff);
    }
  }
  Serial.printf("Read time: %d\n", micros() - start_time);
}

void loop() {
  digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
  delay(250);
 }
```
Does not crash the chip, but shows a failure on the last write or read:
```
EEPROM Test
EEPROM length: 512 Max Test: 256
Write time: 82917
Start read back
Read mismatch: 255 fe!=ff
Read time: 8410
```
And if I reset the chip to run again:  The system then says the EEPROM has 0 bytes, (from  EEPROM.length()) 
```
EEPROM Test
EEPROM length: 0 Max Test: 0
Write time: 1
Start read back
Read time: 1
```
So believe there still are issues.

OpusK commented:
@KurtE, 

I got your repository, updated lib.a, and tested it.
I did 50 tests (reset and open serial monitor) and everything worked fine.

It is difficult because the situation you mentioned is not reproduced. Your repository does not have lib.a updated, did you ever update lib.a and test the example you wrote?

OpusK commented:
@KurtE,

There seems to be a problem with the EEPROM code.
This code comes from [the ST example](https://www.st.com/content/ccc/resource/technical/document/application_note/ee/ef/d7/87/cb/b7/48/52/CD00165693.pdf/files/CD00165693.pdf/jcr:content/translations/en.CD00165693.pdf), and I'll debug it to see if it's implemented correctly.

KurtE commented:
@OpusK,
I will try it again, maybe trying it on a different OpenCM board and see if that makes a difference. 

I had not yet updated the lib file yet, as I don't use it when I debug.   I am setup like I mentioned in:
https://community.robotsource.org/t/opencm-9-04-how-to-build-using-the-develop-branch-and-make-pull-requests/1474/4

FYI - In that Robotsource one I finally went and figured out why I was getting a command line warning on every file being built... Turned out I needed to rename the link I created in the <arduino folder>/hardware/opencm904  to not be opencm9.04 but instead: OpenCM904  as the build was creating the define:  -DARDUINO_ARCH_OPENCM9.04  on each command line and it compiler does not like that . ... 

So I am recompiling the sources each time, I build (with Arduino caching...) And I have an empty dummy 
lib_f103.a file in the github project, in my case: D:\GitHub\OpenCM9.04\arduino\opencm_arduino\opencm9.04\variants\OpenCM904

Only when I am wanting to get the PR, do I rebuild the one that is using eclipse...  Just uploaded it now.

Again I will retry ...

Ooops looks like the adding of the lib file caused conflict...

KurtE commented:
@OpusK 
FYI - I changed the test app, to allow me to do multiple iterations, without reset...
Nothing major:
```
#define MAX_INDEX 256
#include <EEPROM.h>
void setup() {
  // put your setup code here, to run once:
  while (!Serial) ;
  pinMode(LED_BUILTIN, OUTPUT);
  Serial.begin(115200);
  Serial.println("EEPROM Test");

}

void loop() {
  // lets try writing to EEPROM
  uint16_t max_index = EEPROM.length();
  if (max_index > MAX_INDEX) max_index = MAX_INDEX;
  Serial.printf("EEPROM length: %d Max Test: %d\n", EEPROM.length(), max_index);
  uint32_t start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    EEPROM.write(i, i & 0xff);
  }
  Serial.printf("Write time: %d\n", micros() - start_time);
  Serial.println("Start read back");
  start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    uint8_t b = EEPROM.read(i);
    if (b != (i & 0xff)) {
      Serial.printf("Read mismatch: %d %x!=%x\n", i, b, i & 0xff);
    }
  }
  Serial.printf("Read time: %d\n", micros() - start_time);
  Serial.println("Press anykey to run again");
  while (!Serial.available()) {
    digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
    delay(250);
  }
  while (Serial.available()) {
    Serial.read();  // discard everything
  }
}
```
I iterated 4 times, each time it failed on the 0xff... I then reset the board and reopened the monitor and And again I received:
```
EEPROM Test
EEPROM length: 0 Max Test: 0
Write time: 2
Start read back
Read time: 1
Press anykey to run again
```
This was with a different OpenCM9.04 board.  So I will also take a look through the document you sent the link to, and see if I notice anything obvious.

Kurt

KurtE commented:
@OpusK 
> @KurtE,
> 
> There seems to be a problem with the EEPROM code.
> This code comes from [the ST example](https://www.st.com/content/ccc/resource/technical/document/application_note/ee/ef/d7/87/cb/b7/48/52/CD00165693.pdf/files/CD00165693.pdf/jcr:content/translations/en.CD00165693.pdf), and I'll debug it to see if it's implemented correctly.

Looking over this document, the first thing I wonder is: What is really the valid number of EEPROM variables?    Maybe I am wrong, but I think it is 254.

If I am understanding this document correctly, we have 2 1K pages setup for the EEPROM and each element takes 4 bytes (2 for virtual address) and 2 for data.... So at first you think 2K/4 = 512 which is what the
#define NB_OF_VAR             (512)

is defined as... But If I understand this correctly, the first 2 bytes of each page, is used to store that status:
```
  /* Get Page0 status */
  pagestatus0 = (*(__IO uint16_t*)PAGE0_BASE_ADDRESS);
  /* Get Page1 status */
  pagestatus1 = (*(__IO uint16_t*)PAGE1_BASE_ADDRESS);
```

Then my quick look through, it appears like the two pages are used to copy data back and forth.   EEPROM.write(), will write 4 bytes to end of current page.  If that page fills up, it then copies the data in current page to other page, where it takes the last written value for each address and puts it into the copy page, except for new address value.  When copy is done it will assume that there is room left in the copy page for new data...  But all of the values I wrote were unique and not the 0xff value, so again now room left... 

So again wonder if eeprom size is 254?

KurtE commented:
Hi again @OpusK,

Some more things I wonder about or suggest reading through the code. 

On EEPROM.write() and what it calls, there is no validation of what logical register you are trying to write. 
That is suppose that we say there are only 254 memory slots valid.  Should the user be able to set registers like (-1, 4000, 500, 2, 10), which might work, but there is nothing that restricts what indexes you pass in.  

Wonder if the code should check if the value < 0 or >= NB_OF_VAR and bail before trying to store the values?

Wonder if it makes sense to try to limit the wear and speed of EEPROM.write(), such that it does not write the value if the value is already set to that value...   That is it would not be hard to change the function EE_VerifyPageFullWriteVariable to do this.   
```
static uint16_t EE_VerifyPageFullWriteVariable(uint16_t VirtAddress, uint16_t Data)
{
  HAL_StatusTypeDef flashstatus = HAL_OK;
  uint16_t validpage = PAGE0;
  uint32_t address = EEPROM_START_ADDRESS, pageendaddress = EEPROM_START_ADDRESS+PAGE_SIZE;
  uint16_t addressvalue = 0x5555;
  uint16_t cur_Data = !Data;  // initialize to make sure it does not match...

  /* Get valid Page for write operation */
  validpage = EE_FindValidPage(WRITE_IN_VALID_PAGE);

  /* Check if there is no valid page */
  if (validpage == NO_VALID_PAGE)
  {
    return  NO_VALID_PAGE;
  }

  /* Get the valid Page start address */
  address = (uint32_t)(EEPROM_START_ADDRESS + (uint32_t)(validpage * PAGE_SIZE));

  /* Get the valid Page end address */
  pageendaddress = (uint32_t)((EEPROM_START_ADDRESS - 1) + (uint32_t)((validpage + 1) * PAGE_SIZE));

  /* Check each active page address starting from begining */
  while (address < pageendaddress)
  {
    /* Verify if address and address+2 contents are 0xFFFFFFFF */
    if ((*(__IO uint32_t*)address) == 0xFFFFFFFF)
    {
      /* Found an empty slot, see if our index is in the list and the last */
      /* Instance has the value we are trying to store, we can bypass this. */
      if (cur_Data == Data)
      {
        return HAL_OK;  // Value is already set to our value
      }
      /* Set variable data */
      flashstatus = HAL_FLASH_Program(FLASH_TYPEPROGRAM_HALFWORD, address, Data);
      /* If program operation was failed, a Flash error code is returned */
      if (flashstatus != HAL_OK)
      {
        return flashstatus;
      }
      /* Set variable virtual address */
      flashstatus = HAL_FLASH_Program(FLASH_TYPEPROGRAM_HALFWORD, address + 2, VirtAddress);
      /* Return program operation status */
      return flashstatus;
    }
    else
    {
      addressvalue = (*(__IO uint16_t*)(address+2));

      /* Compare the read address with the virtual address */
      if (addressvalue == VirtAddress)
      {
        /* Get content of Address-2 which is variable value */
        cur_Data = (*(__IO uint16_t*)(address));

      }
      /* Next address location */
      address = address + 4;
    }
  }

  /* We did not find any open slots, so if we did not find oru value */
  /* Return PAGE_FULL in case the valid page is full */
  return (cur_Data == Data)? HAL_OK : PAGE_FULL;
}
```
Wish I could figure out how to color code the code to show what I added, but it is pretty basic here.

OpusK commented:
Hi, @KurtE 

I am doing various tests including the code you uploaded.
I have found that the transfer from the current library is not working normally (no recovery function) due to the index issues of PAGE0 and PAGE1(It occurred after forcibly executing EE_Format()).

In addition, when I tried to access the last 4 bytes of each page, I found an error and resetting.
In this regard, I think I should try to debug more tomorrow.

Are you working without problems now?

KurtE commented:
Hi @OpusK,

I was sort of waiting to hear from you, on your findings and if you concur, on some of the findings/assumptions, before I did anything else. 

My findings and suggestions, which I will try out include:

a) #define NB_OF_VAR             (512) should be:
#define NB_OF_VAR             (255)
As I believe you only use one page to look at and store values...   Probably could be reworked to store actual values in both pages, but would take some major reworking.   A hint to this is in the function:
uint16_t EE_ReadVariable(uint16_t VirtAddress, uint16_t* Data)
It looks for THE valid page and scans for the address passed in... It does not scan both pages.

So 1024/4 = 256, but first 2 bytes of page store status so can not store variable so ... 255

b) Don't store any address >= NB_OF_VAR - son in this case only addresses: 0-254
So change: 
uint16_t EE_WriteVariable(uint16_t VirtAddress, uint16_t Data)
and add test:
    if(VirtAddress >= NB_OF_VAR ) return (uint16_t)-1; // return anything but PAGE_FULL

c) I like the idea of not doing a write if the current value is that value... Saves wear on the chip plus speeds it up... 

I will make these changes and push it up to github (assuming they appear to work) and let you try these out. 

Kurt

KurtE commented:
@OpusK,

I made the changes mentioned above, plus update lib_f103.a file with it.  So far it appears to be working. 
Note: If you output all 255 valid addresses (0-254), all new write will cause the whole page to be rewritten, that is every write will cause Page0->Page1 or Page1->Page0... Real slow... 

I changed my test to do 64 addresses (more reasonable).  Also hacked the test case that if when waiting for new key to re-run the test, you enter an 'i', it increments the value, being output to each one.  SO will cause the next run to actually write new values.... 

```
#define MAX_INDEX 64
uint8_t inc_value = 0;  // default to just index...
#include <EEPROM.h>
void setup() {
  // put your setup code here, to run once:
  while (!Serial) ;
  pinMode(LED_BUILTIN, OUTPUT);
  Serial.begin(115200);
  Serial.println("EEPROM Test");

}

void loop() {
  // lets try writing to EEPROM
  uint16_t max_index = EEPROM.length();
  if (max_index > MAX_INDEX) max_index = MAX_INDEX;
  Serial.printf("EEPROM length: %d Max Test: %d Inc: %d\n", EEPROM.length(), max_index, inc_value);
  Serial.flush(); // Make sure the output is out
  uint32_t start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    EEPROM.write(i, (i+inc_value) & 0xff);
  }
  Serial.printf("Write time: %d\n", micros() - start_time);
  Serial.println("Start read back");
  start_time = micros();
  for (uint16_t i = 0; i < max_index; i++) {
    uint8_t b = EEPROM.read(i);
    if (b != ((i+inc_value) & 0xff)) {
      Serial.printf("Read mismatch: %d %x!=%x\n", i, b, (i+inc_value) & 0xff);
    }
  }
  Serial.printf("Read time: %d\n", micros() - start_time);
  Serial.println("Press anykey to run again");
  while (!Serial.available()) {
    digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
    delay(250);
  }
  if (Serial.read() == 'i') inc_value++;    // setup so we can increment all values...
  while (Serial.available()) {
    Serial.read();  // discard everything
  }
}
```
Let me know what you think... 
Note: My update lib_f103.a shows conflict with the one you updated...

KurtE commented:
@OpusK 
Other side comment:  Thought I would do a quick check of the sources versus the OpenCR sources.  
It does not have the problem of PAGE1 (as it is 0x1...)

However, was wondering, about some of the defines, in particular:

    #define PAGE_SIZE             (uint32_t)0x8000  /* Page size */

Or 32768 bytes, now if each entry in that page takes 4 bytes:  as the functions sort of imply like:
EE_VerifyPageFullWriteVariable
Than I would expect that we have space in one page for: 32768/4 - 1 = 8191

But: #define NB_OF_VAR             (4*1024) //4KB

I may be missing something, or simply it only allows half the range of values to help minimize how many times the pages have to be swapped.... Again if this is the case would suggest adding test to not allow writes to address >= NB_OF_VAR...

KurtE commented:
Again not sure if it helps or not, but did another quick and dirty EEPROM test program to dump contents, both what EEPROM.read() shows, plus the raw data of the two Pages.

```
#include <EEPROM.h>

#define ADDR_FLASH_PAGE_126   ((uint32_t)0x0801F800) /* Base @ of Page 126, 1 Kbytes */
#define ADDR_FLASH_PAGE_127   ((uint32_t)0x0801FC00) /* Base @ of Page 127, 1 Kbytes */

/* Define the size of the sectors to be used */
#define PAGE_SIZE               (uint32_t)FLASH_PAGE_SIZE  /* Page size */

/* EEPROM start address in Flash */
#define EEPROM_START_ADDRESS  ((uint32_t)ADDR_FLASH_PAGE_126) /* EEPROM emulation start address */

/* Pages 0 and 1 base and end addresses */
#define PAGE0_BASE_ADDRESS    ((uint32_t)(EEPROM_START_ADDRESS + 0x0000))
#define PAGE0_END_ADDRESS     ((uint32_t)(EEPROM_START_ADDRESS + (PAGE_SIZE - 1)))
#define PAGE0_ID               ADDR_FLASH_PAGE_126

#define PAGE1_BASE_ADDRESS    ((uint32_t)(EEPROM_START_ADDRESS + PAGE_SIZE))
#define PAGE1_END_ADDRESS     ((uint32_t)(EEPROM_START_ADDRESS + PAGE_SIZE + PAGE_SIZE - 1))
#define PAGE1_ID               ADDR_FLASH_PAGE_127
extern void  dumpMemory(uint32_t start_address, uint32_t end_address);

void setup() {
  while(!Serial) ;
  Serial.begin(115200);

  pinMode(LED_BUILTIN, OUTPUT);

  // Lets print out all of the Possible EEPROM values.
  uint16_t eeprom_size = EEPROM.length();
  Serial.printf("EEPROM length: %d\nContents:\n", eeprom_size);

  for (uint16_t i=0; i < eeprom_size; i++) {
    Serial.printf("%x:%x ", i, EEPROM.read(i));

    if ((i & 0xf) == 0xf) Serial.println();
  }
  Serial.println("\nRaw Data");

  // Now lets see if we can dump the 
  Serial.println("EEPROM Page 0");
  dumpMemory(PAGE0_BASE_ADDRESS, PAGE0_END_ADDRESS);

  Serial.println("EEPROM Page 1");
  dumpMemory(PAGE1_BASE_ADDRESS, PAGE1_END_ADDRESS);
}

void  dumpMemory(uint32_t start_address, uint32_t end_address) {
  __IO uint16_t *addr = (__IO uint16_t*)start_address;
  uint8_t count = 0; 
  while (addr < (__IO uint16_t *)end_address) {
    Serial.printf("%x:%x ", addr[0], addr[1]);
    addr += 2;
    count++;
    if (count == 16) {
      Serial.println();
      count = 0;
    }
  }
  Serial.println();
}


void loop() {
  // put your main code here, to run repeatedly:
  digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
  delay(250);
}
```
I was curious if the firmware recovery actually put in a few values... Looks like it did not...
```
EEPROM length: 255
Contents:
0:0 1:0 2:0 3:0 4:0 5:0 6:0 7:0 8:0 9:0 a:0 b:0 c:0 d:0 e:0 f:0 
10:0 11:0 12:0 13:0 14:0 15:0 16:0 17:0 18:0 19:0 1a:0 1b:0 1c:0 1d:0 1e:0 1f:0 
20:0 21:0 22:0 23:0 24:0 25:0 26:0 27:0 28:0 29:0 2a:0 2b:0 2c:0 2d:0 2e:0 2f:0 
30:0 31:0 32:0 33:0 34:0 35:0 36:0 37:0 38:0 39:0 3a:0 3b:0 3c:0 3d:0 3e:0 3f:0 
40:0 41:0 42:0 43:0 44:0 45:0 46:0 47:0 48:0 49:0 4a:0 4b:0 4c:0 4d:0 4e:0 4f:0 
50:0 51:0 52:0 53:0 54:0 55:0 56:0 57:0 58:0 59:0 5a:0 5b:0 5c:0 5d:0 5e:0 5f:0 
60:0 61:0 62:0 63:0 64:0 65:0 66:0 67:0 68:0 69:0 6a:0 6b:0 6c:0 6d:0 6e:0 6f:0 
70:0 71:0 72:0 73:0 74:0 75:0 76:0 77:0 78:0 79:0 7a:0 7b:0 7c:0 7d:0 7e:0 7f:0 
80:0 81:0 82:0 83:0 84:0 85:0 86:0 87:0 88:0 89:0 8a:0 8b:0 8c:0 8d:0 8e:0 8f:0 
90:0 91:0 92:0 93:0 94:0 95:0 96:0 97:0 98:0 99:0 9a:0 9b:0 9c:0 9d:0 9e:0 9f:0 
a0:0 a1:0 a2:0 a3:0 a4:0 a5:0 a6:0 a7:0 a8:0 a9:0 aa:0 ab:0 ac:0 ad:0 ae:0 af:0 
b0:0 b1:0 b2:0 b3:0 b4:0 b5:0 b6:0 b7:0 b8:0 b9:0 ba:0 bb:0 bc:0 bd:0 be:0 bf:0 
c0:0 c1:0 c2:0 c3:0 c4:0 c5:0 c6:0 c7:0 c8:0 c9:0 ca:0 cb:0 cc:0 cd:0 ce:0 cf:0 
d0:0 d1:0 d2:0 d3:0 d4:0 d5:0 d6:0 d7:0 d8:0 d9:0 da:0 db:0 dc:0 dd:0 de:0 df:0 
e0:0 e1:0 e2:0 e3:0 e4:0 e5:0 e6:0 e7:0 e8:0 e9:0 ea:0 eb:0 ec:0 ed:0 ee:0 ef:0 
f0:0 f1:0 f2:0 f3:0 f4:0 f5:0 f6:0 f7:0 f8:0 f9:0 fa:0 fb:0 fc:0 fd:0 fe:0 
Raw Data
EEPROM Page 0
0:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
... all rest was 
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 

EEPROM Page 1
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
... all rest was 
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
```
If I run and incremented, as to write a couple values per address:
```
EEPROM length: 255
Contents:
0:1 1:2 2:3 3:4 4:5 5:6 6:7 7:8 8:9 9:a a:b b:c c:d d:e e:f f:10 
10:11 11:12 12:13 13:14 14:15 15:16 16:17 17:18 18:19 19:1a 1a:1b 1b:1c 1c:1d 1d:1e 1e:1f 1f:20 
20:21 21:22 22:23 23:24 24:25 25:26 26:27 27:28 28:29 29:2a 2a:2b 2b:2c 2c:2d 2d:2e 2e:2f 2f:30 
30:31 31:32 32:33 33:34 34:35 35:36 36:37 37:38 38:39 39:3a 3a:3b 3b:3c 3c:3d 3d:3e 3e:3f 3f:40 
40:0 41:0 42:0 43:0 44:0 45:0 46:0 47:0 48:0 49:0 4a:0 4b:0 4c:0 4d:0 4e:0 4f:0 
50:0 51:0 52:0 53:0 54:0 55:0 56:0 57:0 58:0 59:0 5a:0 5b:0 5c:0 5d:0 5e:0 5f:0 
60:0 61:0 62:0 63:0 64:0 65:0 66:0 67:0 68:0 69:0 6a:0 6b:0 6c:0 6d:0 6e:0 6f:0 
70:0 71:0 72:0 73:0 74:0 75:0 76:0 77:0 78:0 79:0 7a:0 7b:0 7c:0 7d:0 7e:0 7f:0 
80:0 81:0 82:0 83:0 84:0 85:0 86:0 87:0 88:0 89:0 8a:0 8b:0 8c:0 8d:0 8e:0 8f:0 
90:0 91:0 92:0 93:0 94:0 95:0 96:0 97:0 98:0 99:0 9a:0 9b:0 9c:0 9d:0 9e:0 9f:0 
a0:0 a1:0 a2:0 a3:0 a4:0 a5:0 a6:0 a7:0 a8:0 a9:0 aa:0 ab:0 ac:0 ad:0 ae:0 af:0 
b0:0 b1:0 b2:0 b3:0 b4:0 b5:0 b6:0 b7:0 b8:0 b9:0 ba:0 bb:0 bc:0 bd:0 be:0 bf:0 
c0:0 c1:0 c2:0 c3:0 c4:0 c5:0 c6:0 c7:0 c8:0 c9:0 ca:0 cb:0 cc:0 cd:0 ce:0 cf:0 
d0:0 d1:0 d2:0 d3:0 d4:0 d5:0 d6:0 d7:0 d8:0 d9:0 da:0 db:0 dc:0 dd:0 de:0 df:0 
e0:0 e1:0 e2:0 e3:0 e4:0 e5:0 e6:0 e7:0 e8:0 e9:0 ea:0 eb:0 ec:0 ed:0 ee:0 ef:0 
f0:0 f1:0 f2:0 f3:0 f4:0 f5:0 f6:0 f7:0 f8:0 f9:0 fa:0 fb:0 fc:0 fd:0 fe:0 
Raw Data
EEPROM Page 0
0:ffff 0:0 1:1 2:2 3:3 4:4 5:5 6:6 7:7 8:8 9:9 a:a b:b c:c d:d e:e 
f:f 10:10 11:11 12:12 13:13 14:14 15:15 16:16 17:17 18:18 19:19 1a:1a 1b:1b 1c:1c 1d:1d 1e:1e 
1f:1f 20:20 21:21 22:22 23:23 24:24 25:25 26:26 27:27 28:28 29:29 2a:2a 2b:2b 2c:2c 2d:2d 2e:2e 
2f:2f 30:30 31:31 32:32 33:33 34:34 35:35 36:36 37:37 38:38 39:39 3a:3a 3b:3b 3c:3c 3d:3d 3e:3e 
3f:3f 1:0 2:1 3:2 4:3 5:4 6:5 7:6 8:7 9:8 a:9 b:a c:b d:c e:d f:e 
10:f 11:10 12:11 13:12 14:13 15:14 16:15 17:16 18:17 19:18 1a:19 1b:1a 1c:1b 1d:1c 1e:1d 1f:1e 
20:1f 21:20 22:21 23:22 24:23 25:24 26:25 27:26 28:27 29:28 2a:29 2b:2a 2c:2b 2d:2c 2e:2d 2f:2e 
30:2f 31:30 32:31 33:32 34:33 35:34 36:35 37:36 38:37 39:38 3a:39 3b:3a 3c:3b 3d:3c 3e:3d 3f:3e 
40:3f ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
...
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 

EEPROM Page 1
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
...
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 


```

If I run my test a few more times, where I do increments and the like enough times.  I got it to something like:
```
EEPROM length: 255
Contents:
0:2 1:3 2:4 3:5 4:6 5:7 6:8 7:9 8:a 9:b a:c b:d c:e d:f e:10 f:11 
10:12 11:13 12:14 13:15 14:16 15:17 16:18 17:19 18:1a 19:1b 1a:1c 1b:1d 1c:1e 1d:1f 1e:20 1f:21 
20:22 21:23 22:24 23:25 24:26 25:27 26:28 27:29 28:2a 29:2b 2a:2c 2b:2d 2c:2e 2d:2f 2e:30 2f:31 
30:32 31:33 32:34 33:35 34:36 35:37 36:38 37:39 38:3a 39:3b 3a:3c 3b:3d 3c:3e 3d:3f 3e:40 3f:41 
40:0 41:0 42:0 43:0 44:0 45:0 46:0 47:0 48:0 49:0 4a:0 4b:0 4c:0 4d:0 4e:0 4f:0 
50:0 51:0 52:0 53:0 54:0 55:0 56:0 57:0 58:0 59:0 5a:0 5b:0 5c:0 5d:0 5e:0 5f:0 
60:0 61:0 62:0 63:0 64:0 65:0 66:0 67:0 68:0 69:0 6a:0 6b:0 6c:0 6d:0 6e:0 6f:0 
70:0 71:0 72:0 73:0 74:0 75:0 76:0 77:0 78:0 79:0 7a:0 7b:0 7c:0 7d:0 7e:0 7f:0 
80:0 81:0 82:0 83:0 84:0 85:0 86:0 87:0 88:0 89:0 8a:0 8b:0 8c:0 8d:0 8e:0 8f:0 
90:0 91:0 92:0 93:0 94:0 95:0 96:0 97:0 98:0 99:0 9a:0 9b:0 9c:0 9d:0 9e:0 9f:0 
a0:0 a1:0 a2:0 a3:0 a4:0 a5:0 a6:0 a7:0 a8:0 a9:0 aa:0 ab:0 ac:0 ad:0 ae:0 af:0 
b0:0 b1:0 b2:0 b3:0 b4:0 b5:0 b6:0 b7:0 b8:0 b9:0 ba:0 bb:0 bc:0 bd:0 be:0 bf:0 
c0:0 c1:0 c2:0 c3:0 c4:0 c5:0 c6:0 c7:0 c8:0 c9:0 ca:0 cb:0 cc:0 cd:0 ce:0 cf:0 
d0:0 d1:0 d2:0 d3:0 d4:0 d5:0 d6:0 d7:0 d8:0 d9:0 da:0 db:0 dc:0 dd:0 de:0 df:0 
e0:0 e1:0 e2:0 e3:0 e4:0 e5:0 e6:0 e7:0 e8:0 e9:0 ea:0 eb:0 ec:0 ed:0 ee:0 ef:0 
f0:0 f1:0 f2:0 f3:0 f4:0 f5:0 f6:0 f7:0 f8:0 f9:0 fa:0 fb:0 fc:0 fd:0 fe:0 
Raw Data
EEPROM Page 0
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
...
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 

EEPROM Page 1
0:ffff 40:3f 1:0 2:1 3:2 4:3 5:4 6:5 7:6 8:7 9:8 a:9 b:a c:b d:c e:d 
f:e 10:f 11:10 12:11 13:12 14:13 15:14 16:15 17:16 18:17 19:18 1a:19 1b:1a 1c:1b 1d:1c 1e:1d 
1f:1e 20:1f 21:20 22:21 23:22 24:23 25:24 26:25 27:26 28:27 29:28 2a:29 2b:2a 2c:2b 2d:2c 2e:2d 
2f:2e 30:2f 31:30 32:31 33:32 34:33 35:34 36:35 37:36 38:37 39:38 3a:39 3b:3a 3c:3b 3d:3c 3e:3d 
3f:3e 2:0 3:1 4:2 5:3 6:4 7:5 8:6 9:7 a:8 b:9 c:a d:b e:c f:d 10:e 
11:f 12:10 13:11 14:12 15:13 16:14 17:15 18:16 19:17 1a:18 1b:19 1c:1a 1d:1b 1e:1c 1f:1d 20:1e 
21:1f 22:20 23:21 24:22 25:23 26:24 27:25 28:26 29:27 2a:28 2b:29 2c:2a 2d:2b 2e:2c 2f:2d 30:2e 
31:2f 32:30 33:31 34:32 35:33 36:34 37:35 38:36 39:37 3a:38 3b:39 3c:3a 3d:3b 3e:3c 3f:3d 40:3e 
41:3f ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
...
ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff ffff:ffff 
```

OpusK commented:
Hi, @KurtE, 

My problem was caused by the problem you were experiencing for the first time.
As you say, the size should be a maximum page size - 1.
As shown in the documentation, the first 4 bytes of the EEPROM page are used for page status information.

So, when I write to 256 size, I wrote to the status area of the next page, so there was a problem after that.

And, I think your method of reducing unnecessary write while comparing data is very good!

And, I will check about OpenCR's EEPROM code.

KurtE commented:
Hi, @OpusK,

I think OpenCR is probably fine.  

> As you say, the size should be a maximum page size - 1.

Actually the size should be <= (Page Size)/(Size of Item) - 1 ...

That is if you set all values 0 to (NB_OF_VAR-1)  Then after that Each EEPROM.write will cause the whole page to have to rewrite...   So with OpenCR Where MAX could be 8K and is currently 4K, then it will take another 4K EEPROM.writes before the page needs to be compressed and rewritten, which is pretty good. 

So an interesting idea for OpenCM is.  Suppose you restrict the NB_OF_VAR to < 256 and also as EEPROM.write only writes 1 byte elements.   The driver file could be more majorly modified to only use 2 bytes at a time instead of 4.  Index of 0xff is reserved...   

It would then imply if all 255 values were used, you would still have 256 writes before the page would need to be compressed and rewritten...  Just a thought.

OpusK commented:
Hi, @KurtE ,

> So an interesting idea for OpenCM is. Suppose you restrict the NB_OF_VAR to < 256 and also as EEPROM.write only writes 1 byte elements. The driver file could be more majorly modified to only use 2 bytes at a time instead of 4. Index of 0xff is reserved...

>It would then imply if all 255 values were used, you would still have 256 writes before the page would need to be compressed and rewritten... Just a thought.

Is it a way to limit virtual address and data to 1 byte?

Since the official EEPROM library of Arduino is defined as storing 1 byte value as shown below, it seems to be meaningful. (And if that brings speed!)

[EEPROM.write()](https://www.arduino.cc/en/Reference/EEPROMWrite)
> value: the value to write, from 0 to 255 (byte)

KurtE commented:
Hi @OpusK,

Yes the official EEPROM only stores one byte per entry, so yes that is what was driving me to maybe convert to to byte entries. 

I am testing a version now up at: https://github.com/KurtE/OpenCM9.04/tree/eeprom-two-byte-entries
The current version up there is not currently configured to try to maintain the data that was already there.

I have code in there that is currently #ifdef out, that the init code tries to detect that the EEPROM is in the current 4 byte outputs and then convert the data over to 2 byte entries...  But was not working yet. 

I think I know the issue in that I was trying to give a new Page marker for 
#define VALID_PAGE            ((uint16_t)0x0000)     /* Page containing valid data  -2 bytes per entry*/

And define a new type of VALID_PAGE4 that was 0

So if at init it saw 0000 it did the translate.  Problem is I chose VALID_PAGE of 0x0101 And that does not work.  My guess is that if you do a write to this storage you can only turn bits off.  That is why 
FFFF->EEEE->0000 works.  But FFF->EEE->0101 does not.. So I will try again with different marker value: 
maybe 8888  or maybe 4242...

Will be trying this out again in the next few hours.

KurtE commented:
@OpusK,

I have a version of the 2 byte EEPROM settings, that also tries to recover the 4 byte version data and translate it into new 2 byte version.

It is in #61 which does not have conflict of lib_f103.a file... 

You might try it out and see what you think.
